[SERVICE-280] Make tabbing respect resize constraints

tomrobertsOF opened the pull request:
- Resize constraints are now propagated to all windows in the tabGroup. 
- The tabGroup has its own constraints which is the union of the constraints of all tabs.
- When tabbing by window drop or tearout the constraints will now be checked. If incompatible, an 'invalid' preview will be shown.

MichaelMCoates commented:
Quick UI Comment: 
If you try to tab and get the red banner, then try to snap, you see a quick flash of the red banner before it turns purple. The opposite situation isn't as obvious (flash of purple, then to red). Do we have an easy solution for this? Changing the color back and forth on drag on/off might get slow/weird.

tomrobertsOF commented:
@MichaelMCoates, that's been an issue for a while, it's just more apparent with this change since previously the only way to get the red bar was with incompatible tabStrips. There's a story in our cleanup epic to deal with it (SERVICE-287).

tomrobertsOF commented on src/provider/model/DesktopTabGroup.ts:
```diff
@@ -588,4 +595,45 @@ export class DesktopTabGroup implements DesktopEntity {
         ]);
     }
+
+    private async updateGroupConstraints(): Promise<void> {
```
So this gets called any time a window is added to or removed from a tabGroup. The method both calculates the new constraints and applies them to the tabGroup itself and all its windows.

May be worth talking through it face-to-face and seeing if there is a better way

MichaelMCoates replied:
How many times does it fire when you're dragging from one tab group to another?

tomrobertsOF replied:
Yea, it would fire once on the source group when the tab leaves and once on the target group when the tab is added

MichaelMCoates commented on src/provider/model/DesktopTabGroup.ts:
```diff
@@ -588,4 +595,45 @@ export class DesktopTabGroup implements DesktopEntity {
+            let orientation: keyof typeof result;
+            for (orientation in result) {
+                if (result.hasOwnProperty(orientation)) {
+                    const tabConstratins = tab.applicationState.resizeConstraints[orientation];
```
`tabConstraints`

MichaelMCoates commented on src/provider/model/DesktopWindow.ts:
```diff
@@ -107,7 +107,12 @@ enum ActionOrigin {
 
-type OpenFinWindowEventHandler = <K extends keyof fin.OpenFinWindowEventMap>(event: fin.OpenFinWindowEventMap[K]) => void;
+// VScode has issues with syntax painting/intellisense when there are generic function type variables in a file.
+// (specifically it doesn't reallize the type argument has ended, so thinks everything below is still a type)
```
`realize`

tomrobertsOF commented on src/provider/model/DesktopTabGroup.ts:
```diff
@@ -334,22 +337,22 @@ export class DesktopTabGroup implements DesktopEntity {
                     const tabStripHalfSize: Point = this._window.currentState.halfSize;
                     const state: EntityState = tab.currentState;
                     const center: Point = {x: state.center.x, y: state.center.y - tabStripHalfSize.y};
                     const halfSize: Point = {x: state.halfSize.x, y: state.halfSize.y + tabStripHalfSize.y};
```
Which part? The resize constraints part just restores the value that the app had before being tabbed, so nothing to do with window bounds.

pjbroadbent replied:
Any bounds in "`center`/`halfSize`" format will always be normalised in a way where we don't need to be concerned about if the window has a frame or is win10. The `applyProperties` call below will set both the frame and the bounds of the window within the same call, but `_currentState.frame` is updated before the `setBounds` call, so the window knows whether it needs to add in the padding or not when converting `center`/`halfSize` into `l`/`t`/`r`/`b`.

Both the `setBounds` and `updateOptions` are async, so could in theory complete in either order - but it should always end up with the window in the correct position.

MichaelMCoates replied:
👍

pjbroadbent commented on src/provider/APIHandler.ts:
```diff
@@ -285,9 +285,26 @@ export class APIHandler {
+        // Previous client version had no payload. To avoid breaking changes, we
+        // default to the active tab if no window is specified.
+        if (!payload.window) {
+            group = model.getTabGroup(`${source.uuid}/${source.name}`);
```
`model.getId` can be used to generate these strings.

tomrobertsOF replied:
Ah, cool.

pjbroadbent commented on src/provider/model/DesktopWindow.ts:
```diff
@@ -681,7 +691,11 @@ export class DesktopWindow implements DesktopEntity {
+                    if (typeof delta[property] === 'object') {
+                        Object.assign(this._applicationState[property], delta[property]);
+                    } else {
+                        this.applicationState[property] = delta[property]!;
```
Can use the private `_applicationState` here

tomrobertsOF replied:
Gah! I thought I'd caught them all, but apparently not. Fixing now

MichaelMCoates commented:
Needs a rebase
